File: data/readable_format/fortran/file1163.f90
PROGRAM TestSla_DC62S
  IMPLICIT NONE
  DOUBLE PRECISION V(6)
  DOUBLE PRECISION A, B, R, AD, BD, RD

  ! Test input
  V(1) = 1.0D0
  V(2) = 2.0D0
  V(3) = 3.0D0
  V(4) = 0.1D0
  V(5) = 0.2D0
  V(6) = 0.3D0

  CALL sla_DC62S(V, A, B, R, AD, BD, RD)

  PRINT *, "A: ", A
  PRINT *, "B: ", B
  PRINT *, "R: ", R
  PRINT *, "AD: ", AD
  PRINT *, "BD: ", BD
  PRINT *, "RD: ", RD

CONTAINS

  SUBROUTINE sla_DC62S (V, A, B, R, AD, BD, RD)
    IMPLICIT NONE
    DOUBLE PRECISION V(6),A,B,R,AD,BD,RD
    DOUBLE PRECISION X,Y,Z,XD,YD,ZD,RXY2,RXY,R2,XYP

    X=V(1)
    Y=V(2)
    Z=V(3)
    XD=V(4)
    YD=V(5)
    ZD=V(6)

    RXY2=X*X+Y*Y
    R2=RXY2+Z*Z

    IF (R2.EQ.0D0) THEN
       X=XD
       Y=YD
       Z=ZD
       RXY2=X*X+Y*Y
       R2=RXY2+Z*Z
    END IF

    RXY=SQRT(RXY2)
    XYP=X*XD+Y*YD
    IF (RXY2.NE.0D0) THEN
       A=ATAN2(Y,X)
       B=ATAN2(Z,RXY)
       AD=(X*YD-Y*XD)/RXY2
       BD=(ZD*RXY2-Z*XYP)/(R2*RXY)
    ELSE
       A=0D0
       IF (Z.NE.0D0) THEN
          B=ATAN2(Z,RXY)
       ELSE
          B=0D0
       END IF
       AD=0D0
       BD=0D0
    END IF
    R=SQRT(R2)
    IF (R.NE.0D0) THEN
       RD=(XYP+Z*ZD)/R
    ELSE
       RD=0D0
    END IF

  END SUBROUTINE sla_DC62S

END PROGRAM TestSla_DC62S

File: data/readable_format/cpp/file1163.cpp
#include <iostream>
#include <cmath>
#include <iomanip>

// Define the sla_DC62S function directly in this file.
void sla_DC62S(const double V[6], double& A, double& B, double& R, double& AD, double& BD, double& RD) {
    double X, Y, Z, XD, YD, ZD, RXY2, RXY, R2, XYP;

    X = V[0];
    Y = V[1];
    Z = V[2];
    XD = V[3];
    YD = V[4];
    ZD = V[5];

    RXY2 = X * X + Y * Y;
    R2 = RXY2 + Z * Z;

    if (R2 == 0.0) {
        X = XD;
        Y = YD;
        Z = ZD;
        RXY2 = X * X + Y * Y;
        R2 = RXY2 + Z * Z;
    }

    RXY = sqrt(RXY2);
    XYP = X * XD + Y * YD;
    if (RXY2 != 0.0) {
        A = atan2(Y, X);
        B = atan2(Z, RXY);
        AD = (X * YD - Y * XD) / RXY2;
        BD = (ZD * RXY2 - Z * XYP) / (R2 * RXY);
    } else {
        A = 0.0;
        if (Z != 0.0) {
            B = atan2(Z, RXY);
        } else {
            B = 0.0;
        }
        AD = 0.0;
        BD = 0.0;
    }

    R = sqrt(R2);
    if (R != 0.0) {
        RD = (XYP + Z * ZD) / R;
    } else {
        RD = 0.0;
    }
}

// Main function to test sla_DC62S
int main() {
    double V[6] = {1.0, 2.0, 3.0, 0.1, 0.2, 0.3};
    double A, B, R, AD, BD, RD;

    sla_DC62S(V, A, B, R, AD, BD, RD);

    std::cout << std::fixed << std::setprecision(6); // Set precision for floating-point output
    std::cout << "A: " << A << std::endl;
    std::cout << "B: " << B << std::endl;
    std::cout << "R: " << R << std::endl;
    std::cout << "AD: " << AD << std::endl;
    std::cout << "BD: " << BD << std::endl;
    std::cout << "RD: " << RD << std::endl;

    return 0;
}